hwcomposer/wayland-hwc.cpp

```diff
@@ -460,47 +262,153 @@ struct wl_shell_surface_listener shell_surface_listener = {
+            return BufferTransform::Flip_Rot_90;
+        default:
+            ALOGE("Invalid hwc_transform: %" PRIu32, hwc_transform);
+            abort();
+    }
+}
+
+void surface_context::attach_buffer(buffer& buf) {
+    wl_surface_attach(surface, buf.wl_buffer, 0, 0);
+}
+
+void surface_context::damage_surface(int32_t x, int32_t y, int32_t width, int32_t height) {
+    wl_surface_damage(surface, x, y, width, height);
+}
+
+void surface_context::set_buffer_transform(BufferTransform transform) {
+    wl_surface_set_buffer_transform(surface, static_cast<int32_t>(transform));
+}
+
+void surface_context::set_buffer_scale(double d_scale) {
+    assert(!viewport);
+
+    // Usually with no viewporter the scale is guaranteed to be integer
+    // When supports_cursor_viewport == false, this might not be the case
+    // thus use ceil anyway
+    int32_t scale = static_cast<int32_t>(ceil(d_scale));
+    wl_surface_set_buffer_scale(surface, scale);
+}
+
+void surface_context::set_crop(hwc_frect_t crop) {
+    assert(viewport);
+
+    float x = fmax(0, crop.left);
+    float y = fmax(0, crop.top);
+    float width = fmax(1, crop.right - crop.left);
+    float height = fmax(1, crop.bottom - crop.top);
+
+    wp_viewport_set_source(viewport,
+                           wl_fixed_from_double(x),
+                           wl_fixed_from_double(y),
+                           wl_fixed_from_double(width),
+                           wl_fixed_from_double(height));
+}
+
+void surface_context::set_display_frame(hwc_rect_t rect, double scale) {
+    assert(viewport);
+
+    int width = static_cast<int>(ceil((rect.right - rect.left) / scale));
+    int height = static_cast<int>(ceil((rect.bottom - rect.top) / scale));
+
+    wp_viewport_set_destination(viewport,
+                                std::max(1, width),
+                                std::max(1, height));
+}
+
+void window::reset_per_set_state() {
+    lastLayer = 0;
+    last_layer_buffer = nullptr;
+    if (input_region) {
+        wl_region_subtract(input_region, 0, 0, INT_MAX, INT_MAX);
+    }
+}
+
+void window::set_maximize(bool enabled) {
+    if (xdg_toplevel) {
+        if (enabled) {
+            xdg_toplevel_set_maximized(xdg_toplevel);
+        } else {
+            xdg_toplevel_unset_maximized(xdg_toplevel);
+        }
+    } else {
+        assert(shell_surface);
+        if (enabled) {
+            wl_shell_surface_set_maximized(shell_surface, display->output);
+        } else {
+            ALOGW("wl_shell_surface: does not support un-maximizing");
+        }
+    }
+}
+
+void window::set_title(const char* title) {
+    if (xdg_toplevel) {
+        xdg_toplevel_set_title(xdg_toplevel, title);
+    } else {
+        assert(shell_surface);
+        wl_shell_surface_set_title(shell_surface, title);
+    }
+}
+
+void window::set_app_id(std::string appID) {
+    if (appID != "Waydroid")
+        appID = "waydroid." + appID;
+    this->appID = std::move(appID);
```

Comment:
This means that now `window->appID` has the `waydroid.` prefix and so cannot be matched against an android package name. See `WaydroidWindow.cpp` for places where we want to match a package name.